src/main/java/jenkins/plugins/zulip/ZulipNotifier.java

```diff
@@ -74,73 +83,170 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
             String destinationTopic = ZulipUtil.expandVariables(build, listener,
                     ZulipUtil.getDefaultValue(configuredTopic, build.getParent().getDisplayName()));
             Zulip zulip = new Zulip(DESCRIPTOR.getUrl(), DESCRIPTOR.getEmail(), DESCRIPTOR.getApiKey());
-            zulip.sendStreamMessage(destinationStream, destinationTopic, message);
+            zulip.sendStreamMessage(destinationStream, destinationTopic, fullMessage);
+
+            if (DESCRIPTOR.isPersonalNotify()) {
+                Map<String, StringBuilder> personalChangeStrings = buildPersonalChangeStrings(changes);
+                for (Map.Entry<String, StringBuilder> personalChangeString : personalChangeStrings.entrySet()) {
+                    String privateMessage = messageHeader.toString();
+                    if (personalChangeString.getValue().length() > 0) {
+                        privateMessage += "\n\n" + personalChangeString.getValue().toString();
+                    }
+                    zulip.sendPrivateMessage(personalChangeString.getKey(), privateMessage);
+                }
+            }
         }
         return true;
     }
 
-    private String getChangeSet(Run<?, ?> build) {
+    private Map<String, StringBuilder> buildPersonalChangeStrings(List<CommitInfo> changes) {
+        Map<String, StringBuilder> personalChangeStrings = new HashMap<String, StringBuilder>();
+        if (changes != null && !changes.isEmpty()) {
+            for (CommitInfo change : changes) {
+                if (!ZulipUtil.isValueSet(change.email)) {
+                    continue;
+                }
+
+                StringBuilder commitMsgs = null;
+                if (personalChangeStrings.containsKey(change.email)) {
+                    commitMsgs = personalChangeStrings.get(change.email);
+                } else {
+                    commitMsgs = new StringBuilder();
+                    commitMsgs.append("Changes since the last build:\n");
+                }
+                commitMsgs.append("\n* ").append(change.message);
+
+                personalChangeStrings.put(change.email, commitMsgs);
+            }
+        }
+        return personalChangeStrings;
+    }
+
+    private String buildCommonChangeString(List<CommitInfo> changes) {
         StringBuilder changeString = new StringBuilder();
+        if (changes == null) {
+            changeString.append("Could not determine changes since the last build.");
+        } else if (!changes.isEmpty()) {
+            // If there seems to be a commit message at all, try to list all the changes.
+            changeString.append("Changes since the last build:\n");
+            for (CommitInfo change : changes) {
+                if (ZulipUtil.isValueSet(change.email)) {
+                    changeString.append("\n* ");
+                    changeString.append(hundsonUrlMesssage("", "`" + change.author + "`", "mailto:" + change.email));
```

Comment:
Since this is a Zulip integration, would't it be better to be able to message the person via Zulip instead of sending him an e-mail?

While related, this is not part of the personal notify feature and should be in a separate commit. Please rebase to keep the history clean.

Reply:
Yes, I was thinking about it but I couldn't find how to mention a user knowing his/her email. Even when Zulip id is the same as id on Jenkins I don't know how to mention that user in a message.

Reply:
Now, this can be resolved as I removed non-related changes from the PR.